src/common/select-objects.js

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+
+  get value () {
+    const { value } = this.state
+
+    if (this.props.multi) {
+      return map(value, value => value.value)
+    }
+
+    return value.value
+  }
+
+  set value (value) {
+    this.setState({
+      value
+    })
+  }
+
+  @autobind
+  _handleChange (value) {
+    const { onChange } = this.props
+
+    this.setState({
+      value
+    }, onChange && (() => { onChange(this.value) }))
+  }
+}
+
+// ===================================================================
+
+const SelectWrapper = ({ container, placeholder }) => {
+  const { props } = container
+
+  return (
+    <Select
+      onChange={container._handleChange}
+      optionRenderer={container._renderOption}
+      options={container.options}
+      placeholder={props.placeholder || placeholder}
+      required={props.required}
+      value={container.state.value}
+      valueRenderer={container._renderValue}
+      multi={props.multi}
+    />
+  )
+}
+
+const injectUpdateFunction = ({ container, update }) => {
+  container.componentWillReceiveProps = (nextProps) => {
+    container.options = update(nextProps.options)
+  }
+  container.options = update(container.props.options)
+}
+
+const _createContainersSelector = (set, objects, containerAttr = '$container') => {
+  const containers = {}
+  const objectsByContainer = {}
+
+  forEach(set, object => {
+    const containerRef = object[containerAttr]
+    containers[containerRef] = objects[containerRef]
+
+    let container = objectsByContainer[containerRef]
+
+    if (!container) {
+      container = objectsByContainer[containerRef] = {}
+    }
+
+    container[object.id] = object
+  })
+
+  return { containers, objectsByContainer }
+}
+
+const createContainersSelector = (containerAttr) =>
+  createCollectionWrapper(createSelector(
+    (state, props) => props.options,
+    objects,
+    (set, objects) => _createContainersSelector(set, objects, containerAttr)
+  ))
+
+// ===================================================================
+
+@connectStore(() => {
+  const getPools = createContainersSelector('$poolId')
+
+  return (state, props) => {
+    return {
+      ...getPools(state, props)
+    }
+  }
+}, { withRef: true })
+export class SelectHost extends GenericSelect {
+  componentWillMount () {
+    const update = options => {
+      const { props } = this
+      let newOptions = []
+
+      forEach(props.containers, (pool, poolId) => {
+        newOptions.push({
+          label: pool.name_label || poolId,
+          disabled: true,
+          type: 'pool'
+        })
+
+        newOptions = newOptions.concat(
+          map(props.objectsByContainer[poolId], (host, hostId) => {
+            return {
+              value: hostId,
+              label: host.name_label || hostId,
+              type: 'host'
+            }
+          })
+        )
+      })
+
+      return newOptions
+    }
+
+    injectUpdateFunction({ container: this, update })
```

Comment:
I would much prefer you declare a `_computeOptions(props)` method and call it from `GenericOptions#componentWillMount()` and `GenericOptions#componentWillReceiveProps()`.